graphix/gflow.py

```diff
@@ -458,3 +574,194 @@ def get_adjacency_matrix(graph):
     adjacency_matrix = nx.to_numpy_array(graph, nodelist=node_list)
     adjacency_matrix = MatGF2(adjacency_matrix.astype(int))
     return adjacency_matrix, node_list
+
+
+def check_flow(
+    graph: nx.Graph,
+    input: set[int],
+    output: set[int],
+    flow: dict[int, set],
+    meas_planes: dict[int, str] = {},
+) -> bool:
+    """Check whether the flow is valid.
+
+    Parameters
+    ----------
+    graph: nx.Graph
+        graph (incl. in and out)
+    flow: dict[int, set]
+        flow function. flow[i] is the set of qubits to be corrected for the measurement of qubit i.
+    meas_planes: dict[int, str](optional)
+        measurement planes for each qubits. meas_planes[i] is the measurement plane for qubit i.
+
+
+    Returns
+    -------
+    valid_flow: bool
+        True if the flow is valid. False otherwise.
+    """
+
+    valid_flow = True
+    non_outputs = set(graph.nodes) - output
+    # if meas_planes is given, check whether all measurement planes are "XY"
+    for node, plane in meas_planes.items():
+        if plane != "XY" or node not in non_outputs:
+            valid_flow = False
+            return valid_flow
+
+    odd_flow = {node: find_odd_neighbor(graph, corrections) for node, corrections in flow.items()}
+
+    try:
+        layers, depth = get_layers_from_flow(flow, odd_flow, input, output)
+    except ValueError:
+        valid_flow = False
+        return valid_flow
+    node_order = []
+    for d in range(depth):
+        node_order.extend(list(layers[d]))
+    adjacency_matrix, node_list = get_adjacency_matrix(graph)
+    permute = [node_list.index(i) for i in node_order]
+    adjacency_matrix.permute_col(permute)
+    adjacency_matrix.permute_row(permute)
+
+    flow_matrix = MatGF2(np.zeros((len(node_order), len(node_order)), dtype=int))
+    for node, corrections in flow.items():
+        for correction in corrections:
+            row = node_order.index(node)
+            col = node_order.index(correction)
+            # check whether v < f(v)
+            if row >= col:
+                valid_flow = False
+                return valid_flow
+            # check whether the flow arrow is edge of the original graph
+            if adjacency_matrix.data[row, col] != 1:
+                valid_flow = False
+                return valid_flow
+            flow_matrix.data[row, col] = 1
+
+    Neighbor_f = flow_matrix @ adjacency_matrix
+    # Neighbor_f = MatGF2(Neighbor_f.data)
+    # check whether v < N(f(v))
+    triu = np.triu(Neighbor_f.data)
+    triu = MatGF2(triu)
+    dif = triu - Neighbor_f
+    valid_flow &= np.count_nonzero(dif.data) == 0
+    return valid_flow
+
+
+def check_gflow(
```

Comment:
`verify_gflow`?

Reply:
I agree with this convention, also what do you think about naming the main algorithm `gflow_finder()` istead of just `gflow` ? Since in essence it is lookinf for a gflow in the presented graph.

Reply:
thanks for the comment! Along your line of argument, I would suggest 'find_gflow'.